staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,306 @@
+	}
+}
+
+// Run starts a goroutine that propagates the shutdown signal,
+// individual delegates are ran as they are created.
+func (b *backend) Run(stopCh <-chan struct{}) error {
+	go func() {
+		<-stopCh
+		b.disableUpdates()
+		b.stopAllDelegates()
+	}()
+	// event handler is added after the stop handler is provisioned, this is to prevent any possible
+	// race condition involving the informer adding sinks before Run is called. This will still happen
+	// before the informers are started
+	b.config.Informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			b.addSink(obj.(*auditregv1alpha1.AuditSink))
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			b.updateSink(oldObj.(*auditregv1alpha1.AuditSink), newObj.(*auditregv1alpha1.AuditSink))
+		},
+		DeleteFunc: func(obj interface{}) {
+			b.deleteSink(obj.(*auditregv1alpha1.AuditSink))
+		},
+	})
+	return nil
+}
+
+// disableUpdates permanently disables the ability to modify delegates
+// this is used to shutdown gracefully
+func (b *backend) disableUpdates() {
+	b.delegateUpdateMutex.Lock()
+}
+
+// stopAllDelegates closes the stopChan for every delegate to enable
+// goroutines to terminate gracefully
+func (b *backend) stopAllDelegates() {
+	for _, d := range b.GetDelegates() {
+		close(d.stopChan)
+	}
+}
+
+// Shutdown calls the shutdown method on all delegates
+func (b *backend) Shutdown() {
+	for _, d := range b.GetDelegates() {
+		d.Shutdown()
+	}
+}
+
+// GetDelegates retrieves current delegates in a safe manner
+func (b *backend) GetDelegates() syncedDelegates {
+	return b.delegates.Load().(syncedDelegates)
+}
+
+// setDelegates sets the current delegates in a safe manner
+func (b *backend) setDelegates(delegates syncedDelegates) {
+	b.delegates.Store(delegates)
+}
+
+// addSink is called by the shared informer when a sink is added
+func (b *backend) addSink(sink *auditregv1alpha1.AuditSink) {
+	b.delegateUpdateMutex.Lock()
+	defer b.delegateUpdateMutex.Unlock()
+	delegates := b.GetDelegates()
+	if _, ok := delegates[sink.UID]; ok {
+		glog.Errorf("Audit sink %q uid: %s already exists, could not readd", sink.Name, sink.UID)
+		return
+	}
+	d, err := b.createAndStartDelegate(sink)
+	if err != nil {
+		msg := fmt.Sprintf("Could not add audit sink %q: %v", sink.Name, err)
+		glog.Error(msg)
+		b.recorder.Event(sink, corev1.EventTypeWarning, "CreateFailed", msg)
+		return
+	}
+	delegates[sink.UID] = d
+	b.setDelegates(delegates)
+	glog.V(2).Infof("Added audit sink: %s", sink.Name)
+	glog.V(2).Infof("Current audit sinks: %v", delegates.Names())
+}
+
+// updateSink is called by the shared informer when a sink is updated.
+// The new sink is rebuilt so any changes to the spec may be applied.
+// The new sink must not have the same uid as the previous. The new
+// sink will be started before the old one is shutdown so no events will be lost
+func (b *backend) updateSink(oldSink, newSink *auditregv1alpha1.AuditSink) {
```

Comment:
updateSink can be called for changes on the object that do not affect how it processes events (annotation and label changes, or even occasional informer resyncs can trigger update events). given how expensive stopping an old delegate and starting a new one is, should we make sure the oldSink and newSink differ in ways that affect the delegate we would build before doing all this work?

Reply:
yea I was trying to keep it simple to start, but I can refine the logic here